processor/src/decorators/adv_stack_injectors.rs

```diff
@@ -558,6 +567,190 @@ where
+                            root,
+                            &new_index.depth().into(),
+                            &new_index.value().into(),
+                            new_node.into(),
+                        )?;
+                        new_root
+                    }
+                    None => {
+                        // if the node does not have a lone sibling, we just replace it with an
+                        // empty node
+                        let (_, new_root) = self.advice_provider.update_merkle_node(
+                            root,
+                            &Felt::from(depth),
+                            &index,
+                            empty.into(),
+                        )?;
+                        new_root
+                    }
+                };
+
+                self.advice_provider.push_stack(AdviceSource::Word(leaf_value))?;
+                self.advice_provider.push_stack(AdviceSource::Word(new_root))?;
+
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns info about a lone sibling of a leaf specified by depth and index parameters in the
+    /// Tiered Sparse Merkle tree defined by the specified root. If no lone siblings exist for the
+    /// specified parameters, None is returned.
+    ///
+    /// A lone sibling is defined as a leaf which has a common root with the specified leaf at a
+    /// higher tier such that the subtree starting at this root contains only these two leaves.
+    ///
+    /// In addition to the leaf node itself, this also returns the index of the common root at a
+    /// higher tier.
+    fn find_lone_sibling(
+        &self,
+        root: Word,
+        depth: u8,
+        index: &Felt,
+    ) -> Result<Option<(RpoDigest, NodeIndex)>, ExecutionError> {
+        debug_assert!(matches!(depth, 16 | 32 | 48));
+
+        // if the leaf is on the first tier (depth=16), we don't care about lone siblings as they
+        // cannot be moved to a higher tier.
+        if depth == TieredSmt::TIER_SIZE {
+            return Ok(None);
+        }
+
+        let empty = &EmptySubtreeRoots::empty_hashes(TieredSmt::MAX_DEPTH)[..=depth as usize];
+
+        // get the path to the leaf node
+        let path: Vec<_> = self.advice_provider.get_merkle_path(root, &depth.into(), index)?.into();
+
+        // traverse the path from the leaf up to the root, keeping track of all non-empty nodes;
+        // here we ignore the top 16 depths because lone siblings cannot be moved to a higher tier
+        // from tier at depth 16.
+        let mut non_empty_nodes = BTreeMap::new();
+        for (depth, p_node) in (TieredSmt::TIER_SIZE..=depth).rev().zip(path.iter()) {
```

Comment:
If I'm not mistaken `p_node` is the sibling node on the path.  If this is the case it may be more intuitive to rename this to `sibling_node`.

Reply:
Correct. Renamed to just `sibling`.